### PR TITLE
TaggedMetricRegistry composability

### PR DESCRIPTION
**Pull request by j-baker**
Frequently libraries produce metrics, but the context of usage is not known (e.g. Atlas metrics are coupled to an Atlas client, Spark metrics are tied to a job). In this case, if a caller has multiple metric registries, they can add the child metrics to the parent with a tag, so that they can be distinguished.

### CODE REVIEW

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/DropwizardTaggedMetricSet.java
Comment (j-baker) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/DropwizardTaggedMetricSet.java]: I could avoid the copy here if we think it's important by implementing ForwardingMap.

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/DefaultTaggedMetricRegistry.java
Comment (j-baker) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/DefaultTaggedMetricRegistry.java]: probably not _quite right_, since will want to shove probably two MetricSets in for each AtlasDB, also probably shouldn't be passing in an empty map here. Maybe just a Set of Pair type structure is the right thing here.

Comment (schlosna) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/DefaultTaggedMetricRegistry.java]: Yeah, this feels odd and I worry about someone putting a mutable `Map<String, String>` in as a key and causing all kinds of chaos.

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/DropwizardTaggedMetricSet.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/DropwizardTaggedMetricSet.java]: are we ok to assume all metric names are safe here?

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java:89]: I haven't looked at how painful the API break would be, do you have a sense?

**Review by schlosna — CHANGES_REQUESTED** [review]
Hey James, interested to see how you're planning on using this, maybe sketch out some tests showing intent? I'm wondering if we want to have a single canonical metric registry that other partitions can feed into and auto adjust MetricName to include that partition's tags.

For example if we have service Foo with a single registry, and Atlas clients Bar and Baz, we'd inject a registry that auto adds tag `client=Bar` and `client=Baz` respectively, Atlas adds a counter `qux` and the underlying Foo registry has two counters `qux;client=Bar` and `qux;client=Baz`.

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java:89]: that's not actually an API break; it's just moved into the super interface

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/DropwizardTaggedMetricSet.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/DropwizardTaggedMetricSet.java]: yes - they'll get logged internally in the same way as they currently do. The reason this was called safe name was to make clear that they have to be safe - but that's a baseline requirement.

**Comment by j-baker** [comment]
hey dave - the motivation here is basically: we have multiple atlas clients and want to be able to control how they're reported.

I don't think libraries should be using static metrics. This is for a few reasons, but I've grown conviction in this workflow over time. It's painful for producers because they have to understand the metrics with regards to a lifecycle. For example, with AtlasDB they have to go through the workflows of removing metrics upon failed initialisation tasks, so as to not leave now-unused gauges around, and the complexity of this is nontrivial.

For consumers, it's painful because now the metrics that are being recorded are defined by the producer. For example, AtlasDB made the bad assumption that no one has more than one AtlasDB hanging around, and this was very very painful to change (2000 lines +-), because the metrics needed to be plumbed through (even though it was only like 100 lines of additions overall).

It's also painful for aggregators, since it makes the incremental cost of adding metrics almost zero, which means that tracking which metrics exist and are useful is essentially impossible. For example, AtlasDB by default adds around 1500 metrics. This makes it very difficult to have a coherent metric reporting story (which metrics are even valuable? which should I use?) because they're so numerous.

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java]: API question: should we return either a `boolean` letting consumer know if it was removed, or the removed `TaggedMetricSet`?

**Review by schlosna — COMMENTED** [review]
Hey James, makes sense and agreed AtlasDB should have been injecting the appropriately scoped metrics registry from the beginning.

Had one minor API question before merging.

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java
Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java]: I now return an optional taggedmetricset to be consistent

Comment (j-baker) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java]: with remove()

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/DefaultTaggedMetricRegistry.java
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/DefaultTaggedMetricRegistry.java]: you got checkstyled

```
ParameterNameCheck - com.palantir.tritium.metrics.registry.DefaultTaggedMetricRegistry - checkstyleMain
DefaultTaggedMetricRegistry.java:105: Parameter name 'b' must match pattern '^[a-z][a-zA-Z0-9]+$'.

ParameterNameCheck - com.palantir.tritium.metrics.registry.DefaultTaggedMetricRegistry - checkstyleMain
DefaultTaggedMetricRegistry.java:105: Parameter name 'a' must match pattern '^[a-z][a-zA-Z0-9]+$'.
```

**Review by schlosna — APPROVED** [review]

**Comment by schlosna** [comment]
Approved pending checkstyle fix

**Comment by schlosna** [comment]
@j-baker go ahead and update the description when you're happy to squash & merge

**Comment by jiahuijiang** [comment]
@j-baker Why do we only allow one tag for each metricSet here not a map of tags and their values?

**Pull request merged**
